src/cdk/tree/tree.ts

```diff
@@ -311,6 +571,513 @@ export class CdkTree<T, K = T> implements AfterContentChecked, CollectionViewer,
+  collapseAll(): void {
+    if (this.treeControl) {
+      this.treeControl.collapseAll();
+    } else if (this._expansionModel) {
+      const expansionModel = this._expansionModel;
+      expansionModel.deselect(
+        ...this._flattenedNodes.value.map(child => this._getExpansionKey(child)),
+      );
+    }
+  }
+
+  /** Level accessor, used for compatibility between the old Tree and new Tree */
+  _getLevelAccessor() {
+    return this.treeControl?.getLevel ?? this.levelAccessor;
+  }
+
+  /** Children accessor, used for compatibility between the old Tree and new Tree */
+  _getChildrenAccessor() {
+    return this.treeControl?.getChildren ?? this.childrenAccessor;
+  }
+
+  /**
+   * Gets the direct children of a node; used for compatibility between the old tree and the
+   * new tree.
+   */
+  _getDirectChildren(dataNode: T): Observable<T[]> {
+    const levelAccessor = this._getLevelAccessor();
+    const expansionModel = this._expansionModel ?? this.treeControl?.expansionModel;
+    if (!expansionModel) {
+      return observableOf([]);
+    }
+
+    const key = this._getExpansionKey(dataNode);
+
+    const isExpanded = expansionModel.changed.pipe(
+      switchMap(changes => {
+        if (changes.added.includes(key)) {
+          return observableOf(true);
+        } else if (changes.removed.includes(key)) {
+          return observableOf(false);
+        }
+        return EMPTY;
+      }),
+      startWith(this.isExpanded(dataNode)),
+    );
+
+    if (levelAccessor) {
+      return combineLatest([isExpanded, this._flattenedNodes]).pipe(
+        map(([expanded, flattenedNodes]) => {
+          if (!expanded) {
+            return [];
+          }
+          const startIndex = flattenedNodes.findIndex(node => this._getExpansionKey(node) === key);
+          const level = levelAccessor(dataNode) + 1;
+          const results: T[] = [];
+
+          // Goes through flattened tree nodes in the `flattenedNodes` array, and get all direct
+          // descendants. The level of descendants of a tree node must be equal to the level of the
+          // given tree node + 1.
+          // If we reach a node whose level is equal to the level of the tree node, we hit a sibling.
+          // If we reach a node whose level is greater than the level of the tree node, we hit a
+          // sibling of an ancestor.
+          for (let i = startIndex + 1; i < flattenedNodes.length; i++) {
+            const currentLevel = levelAccessor(flattenedNodes[i]);
+            if (level > currentLevel) {
+              break;
+            }
+            if (level === currentLevel) {
+              results.push(flattenedNodes[i]);
+            }
+          }
+          return results;
+        }),
+      );
+    }
+    const childrenAccessor = this._getChildrenAccessor();
+    if (childrenAccessor) {
+      return coerceObservable(childrenAccessor(dataNode) ?? []);
+    }
+    throw getTreeControlMissingError();
```

Comment:
It would be a bit easier to follow if there was a breakout function like `getDirectChildrenByLevel`